[flutter_tools] Pass --chrome-binary to app hosting Chrome in flutter drive

shindonghwi opened the pull request:
Fixes #171504

## Problem
When running `flutter drive -d chrome --chrome-binary=/path/to/chrome`, the `--chrome-binary` flag was only being passed to WebDriver Chrome, not to the app hosting Chrome instance.

## Solution
Added `webChromeBinary` field to `DebuggingOptions` and passed it through to `ChromiumLauncher.launch()`.

## Changes
- Add `webChromeBinary` field to `DebuggingOptions`
- Add `chromeBinary` parameter to `ChromiumLauncher.launch()`
- Forward `webChromeBinary` in `WebDriverService.start()`
- Add test for custom chrome binary path

## Test
```
flutter drive -d chrome --chrome-binary="/custom/path/chrome" -v
```
Verified that log shows `Will use Chromium executable at /custom/path/chrome`.

shindonghwi committed:
[flutter_tools] Pass --chrome-binary to app hosting Chrome in flutter drive

Fixes #171504

The --chrome-binary flag was only being passed to WebDriver Chrome,
not to the app hosting Chrome instance. This change passes the flag
through DebuggingOptions so both Chrome instances use the specified binary.

Changes:
- Add webChromeBinary field to DebuggingOptions
- Pass chromeBinary parameter to ChromiumLauncher.launch()
- Forward webChromeBinary in WebDriverService.start()
- Add test for custom chrome binary path

shindonghwi committed:
fix: add chromeBinary parameter to test mock classes

Update TestChromiumLauncher.launch() method signature in test files
to include the chromeBinary parameter added to ChromiumLauncher.launch()

bkonyi reviewed (dismissed):
LGTM!

However, it looks like packages/flutter_tools/test/web.shard/chrome_test.dart needs to be formatted.

mdebbar reviewed (commented):
The real bug here is that we are launching two Chrome instances (https://github.com/flutter/flutter/issues/148915), one is the correct one specified by `chrome-binary`, and the other one is unspecified (launched as a side effect of `-d chrome`). This PR is fixing the binary location of the latter, but it's not fixing the root of the issue.

It looks like we already have a mechanism to prevent `-d chrome` from launching a Chrome instance:
https://github.com/flutter/flutter/blob/d7d35fdd9f068b7346de77ee318039ed74518cfa/packages/flutter_tools/lib/src/web/web_device.dart#L138
and `flutter drive` is making an attempt to prevent the extra Chrome instance from launching:
https://github.com/flutter/flutter/blob/d7d35fdd9f068b7346de77ee318039ed74518cfa/packages/flutter_tools/lib/src/commands/drive.dart#L355
but something is failing in between.

<hr>

I recommend going after the root of the issue by fixing the `--no-launch-chrome`, what do you think @shindonghwi ?

shindonghwi committed:
fix: prevent duplicate Chrome launch in flutter drive

flutter drive -d chrome was opening Chrome twice - once from WebDriver
and once from web_device.dart. Added webNoLaunchChrome flag to
DebuggingOptions so we can skip the second launch.

shindonghwi commented:
@mdebbar yeah you're right, I traced through the code.

The platformArgs approach doesn't work because resident_web_runner.dart:363 overwrites it with `{'uri': url}`. So the `no-launch-chrome` key never reaches web_device.dart.

Fixed it by adding `webNoLaunchChrome` to DebuggingOptions and passing it through the whole chain. Also added tests for the flag behavior.

Updated commit pushed.

bkonyi commented:
Just checking in @shindonghwi. Do you have plans to come back to this PR? :)

shindonghwi commented:
@bkonyi Hey, thanks for checking in! Unfortunately, I've been quite busy lately and won't be able to get back to this PR anytime soon. Feel free to take it over or close it if needed. Sorry about that!

bkonyi commented:
Closing in favor of https://github.com/flutter/flutter/pull/185481
